src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -499,235 +415,252 @@ int MPIR_Typerep_wait(MPIR_Typerep_req typerep_req)
-                                      type_size * count);
-                } else {
-                    max_pack_bytes = type_size * count;
-                    if (source_dtp_size == source_dtp_extent) {
-                        mpi_errno =
-                            op_pack((char *) source_buf +
-                                    src_type_stride * accumulated_count, count,
-                                    type, 0,
-                                    (char *) target_buf +
-                                    MPIR_Ptr_to_aint(curr_loc), max_pack_bytes,
-                                    &actual_pack_bytes, op);
-                    } else {
-                        char *ptr = MPL_malloc(source_dtp_extent, MPL_MEM_OTHER);
-                        MPI_Aint unpacked_size;
-                        mpi_errno =
-                            op_pack((char *) source_buf +
-                                    src_type_stride * accumulated_count, count,
-                                    type, 0, ptr, max_pack_bytes, &actual_pack_bytes, op);
-                        MPIR_Typerep_unpack(ptr, source_dtp_size,
-                                            (char *) target_buf + MPIR_Ptr_to_aint(curr_loc), 1,
-                                            source_dtp, 0, &unpacked_size);
-                        MPL_free(ptr);
-                    }
-                }
-                if (curr_len % type_size == 0) {
-                    i++;
-                    if (i != vec_len) {
-                        curr_loc = typerep_vec[i].iov_base;
-                        curr_len = typerep_vec[i].iov_len;
-                    }
-                } else {
-                    curr_loc = (void *) ((char *) curr_loc + type_extent * count);
-                    curr_len -= type_size * count;
-                }
+    bool use_yaksa = MPIR_Typerep_reduce_is_supported(op, source_dtp);
+    int source_is_contig, target_is_contig;
+    MPIR_Datatype_is_contig(source_dtp, &source_is_contig);
+    MPIR_Datatype_is_contig(source_dtp, &target_is_contig);
 
-                accumulated_count += count;
-            }
+    if (use_yaksa) {
+        if (source_is_packed || source_is_contig) {
+            mpi_errno = typerep_op_unpack(source_buf, target_buf, target_count, target_dtp, op);
+        } else if (target_is_contig) {
+            mpi_errno = typerep_op_pack(source_buf, target_buf, source_count, source_dtp, op);
+        } else {
+            MPI_Aint source_dtp_size;
+            MPIR_Datatype_get_size_macro(source_dtp, source_dtp_size);
+            void *src_ptr = MPL_malloc(source_dtp_size * source_count, MPL_MEM_OTHER);
+            MPI_Aint data_sz = source_dtp_size * source_count;
+            MPI_Aint pack_size;
+            MPIR_Typerep_pack(source_buf, source_count, source_dtp, 0, src_ptr, data_sz,
+                              &pack_size);
+            MPIR_Assert(pack_size == data_sz);
```

Comment:
So, we are not generating iovs for yaksa reduction?

Reply:
Yeah, seems it is totally unnecessary. `yaksa_unpack` handles that case perfectly.

Reply:
Yes, but you need to pack the entire data and then call `unpack`. There could be performance implications for large data sizes.

Reply:
There are two cases. With the active message case, the source data is already packed, so yaksa unpack actually skips the extra unpacking step in the old path. With posix, the source data is not packed -- I'm not sure why is that, because unless its IPC, there is no reason to have the source data not packed, right? -- so the case is exactly like `MPIR_Localcopy`, which involves both pack and unpack. We'll need to measure it, but I don't think the IOV path can perform better. This is especially true for the case of pairtypes where the density is really low. Nevertheless, if we need optimization, the optimization should be within Typerep or yaksa, for example, consider provide a noncontig-noncontig general `yaksa_op`. In fact, I think a general noncontig-noncont `yaksa_copy` should also be considered.

Reply:
But before dive into the optimization debate, we should recognize that the RMA pair type accumulation is really a rare/low-priority case, where it is justified to trade for slightly lower performance for better code maintainability.

Reply:
The rest of the changes look good to me.

Reply:
Sounds good. Could you resolve this conversation (or do we need further investigation)? After that I'll rebase and run some more tests and this can be merged soon.